// @mantine/charts/esm/CompositeChart/CompositeChart.mjs
'use client';
import { jsx, jsxs } from 'react/jsx-runtime';
import { useState, createElement } from 'react';
import { Line, Area, Bar, ReferenceLine, ResponsiveContainer, ComposedChart, Legend, XAxis, Label, YAxis, CartesianGrid, Tooltip } from 'recharts';
import { createVarsResolver, getThemeColor, factory, useProps, useMantineTheme, useResolvedStylesApi, useStyles, Box } from '@mantine/core';
import { BarLabel } from '../BarChart/BarChart.mjs';
import { ChartLegend } from '../ChartLegend/ChartLegend.mjs';
import { ChartTooltip } from '../ChartTooltip/ChartTooltip.mjs';
import { PointLabel } from '../PointLabel/PointLabel.mjs';
import classes from '../grid-chart.module.css.mjs';

const defaultProps = {
  withXAxis: true,
  withYAxis: true,
  withTooltip: true,
  tooltipAnimationDuration: 0,
  tickLine: "y",
  strokeDasharray: "5 5",
  gridAxis: "x",
  withDots: true,
  connectNulls: true,
  strokeWidth: 2,
  curveType: "monotone"
};
const varsResolver = createVarsResolver(
  (theme, { textColor, gridColor }) => ({
    root: {
      "--chart-text-color": textColor ? getThemeColor(textColor, theme) : void 0,
      "--chart-grid-color": gridColor ? getThemeColor(gridColor, theme) : void 0
    }
  })
);
const CompositeChart = factory((_props, ref) => {
  const props = useProps("CompositeChart", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    data,
    withLegend,
    legendProps,
    series,
    onMouseLeave,
    dataKey,
    withTooltip,
    withXAxis,
    withYAxis,
    gridAxis,
    tickLine,
    xAxisProps,
    yAxisProps,
    unit,
    tooltipAnimationDuration,
    strokeDasharray,
    gridProps,
    tooltipProps,
    referenceLines,
    withDots,
    dotProps,
    activeDotProps,
    strokeWidth,
    connectNulls,
    curveType,
    dir,
    valueFormatter,
    children,
    lineProps,
    xAxisLabel,
    yAxisLabel,
    withRightYAxis,
    rightYAxisLabel,
    rightYAxisProps,
    withPointLabels,
    areaProps,
    barProps,
    withBarValueLabel,
    minBarSize,
    maxBarWidth,
    composedChartProps,
    ...others
  } = props;
  const theme = useMantineTheme();
  const withXTickLine = gridAxis !== "none" && (tickLine === "x" || tickLine === "xy");
  const withYTickLine = gridAxis !== "none" && (tickLine === "y" || tickLine === "xy");
  const [highlightedArea, setHighlightedArea] = useState(null);
  const shouldHighlight = highlightedArea !== null;
  const handleMouseLeave = (event) => {
    setHighlightedArea(null);
    onMouseLeave?.(event);
  };
  const { resolvedClassNames, resolvedStyles } = useResolvedStylesApi({
    classNames,
    styles,
    props
  });
  const getStyles = useStyles({
    name: "CompositeChart",
    classes,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  const lines = series.map((item) => {
    const color = getThemeColor(item.color, theme);
    const dimmed = shouldHighlight && highlightedArea !== item.name;
    if (item.type === "line") {
      return /* @__PURE__ */ createElement(
        Line,
        {
          ...getStyles("line"),
          key: item.name,
          name: item.name,
          dataKey: item.name,
          dot: withDots ? {
            fillOpacity: dimmed ? 0 : 1,
            strokeOpacity: dimmed ? 0 : 1,
            strokeWidth: 1,
            fill: color,
            stroke: color,
            ...dotProps
          } : false,
          activeDot: withDots ? { fill: "var(--mantine-color-white)", stroke: color, ...activeDotProps } : false,
          fill: color,
          stroke: color,
          strokeWidth,
          isAnimationActive: false,
          fillOpacity: dimmed ? 0 : 1,
          strokeOpacity: dimmed ? 0.5 : 1,
          connectNulls,
          type: curveType,
          strokeDasharray: item.strokeDasharray,
          yAxisId: item.yAxisId || "left",
          label: withPointLabels ? /* @__PURE__ */ jsx(PointLabel, {}) : void 0,
          ...typeof lineProps === "function" ? lineProps(item) : lineProps
        }
      );
    }
    if (item.type === "area") {
      return /* @__PURE__ */ jsx(
        Area,
        {
          ...getStyles("area"),
          name: item.name,
          type: curveType,
          dataKey: item.name,
          fill: color,
          strokeWidth,
          stroke: color,
          isAnimationActive: false,
          connectNulls,
          dot: withDots ? {
            fillOpacity: dimmed ? 0 : 1,
            strokeOpacity: dimmed ? 0 : 1,
            strokeWidth: 1,
            fill: color,
            stroke: color,
            ...dotProps
          } : false,
          activeDot: withDots ? {
            fill: theme.white,
            stroke: color,
            strokeWidth: 2,
            r: 4,
            ...activeDotProps
          } : false,
          fillOpacity: dimmed ? 0 : 0.2,
          strokeOpacity: dimmed ? 0.5 : 1,
          strokeDasharray: item.strokeDasharray,
          yAxisId: item.yAxisId || "left",
          label: withPointLabels ? /* @__PURE__ */ jsx(PointLabel, {}) : void 0,
          ...typeof areaProps === "function" ? areaProps(item) : areaProps
        }
      );
    }
    if (item.type === "bar") {
      return /* @__PURE__ */ createElement(
        Bar,
        {
          ...getStyles("bar"),
          key: item.name,
          name: item.name,
          dataKey: item.name,
          fill: color,
          stroke: color,
          isAnimationActive: false,
          fillOpacity: dimmed ? 0.1 : 1,
          strokeOpacity: dimmed ? 0.2 : 0,
          label: withBarValueLabel ? /* @__PURE__ */ jsx(BarLabel, { valueFormatter }) : void 0,
          yAxisId: item.yAxisId || "left",
          minPointSize: minBarSize,
          ...typeof barProps === "function" ? barProps(item) : barProps
        }
      );
    }
    return null;
  });
  const referenceLinesItems = referenceLines?.map((line, index) => {
    const color = getThemeColor(line.color, theme);
    return /* @__PURE__ */ jsx(
      ReferenceLine,
      {
        stroke: line.color ? color : "var(--chart-grid-color)",
        strokeWidth: 1,
        yAxisId: line.yAxisId || "left",
        ...line,
        label: {
          value: line.label,
          fill: line.color ? color : "currentColor",
          fontSize: 12,
          position: line.labelPosition ?? "insideBottomLeft"
        },
        ...getStyles("referenceLine")
      },
      index
    );
  });
  const sharedYAxisProps = {
    axisLine: false,
    type: "number",
    tickLine: withYTickLine ? { stroke: "currentColor" } : false,
    allowDecimals: true,
    unit,
    tickFormatter: valueFormatter,
    ...getStyles("axis")
  };
  return /* @__PURE__ */ jsx(
    Box,
    {
      ref,
      ...getStyles("root"),
      onMouseLeave: handleMouseLeave,
      dir: dir || "ltr",
      ...others,
      children: /* @__PURE__ */ jsx(ResponsiveContainer, { ...getStyles("container"), children: /* @__PURE__ */ jsxs(
        ComposedChart,
        {
          data,
          maxBarSize: maxBarWidth,
          margin: {
            bottom: xAxisLabel ? 30 : void 0,
            left: yAxisLabel ? 10 : void 0,
            right: yAxisLabel ? 5 : void 0
          },
          ...composedChartProps,
          children: [
            withLegend && /* @__PURE__ */ jsx(
              Legend,
              {
                verticalAlign: "top",
                content: (payload) => /* @__PURE__ */ jsx(
                  ChartLegend,
                  {
                    payload: payload.payload,
                    onHighlight: setHighlightedArea,
                    legendPosition: legendProps?.verticalAlign || "top",
                    classNames: resolvedClassNames,
                    styles: resolvedStyles,
                    series
                  }
                ),
                ...legendProps
              }
            ),
            /* @__PURE__ */ jsxs(
              XAxis,
              {
                hide: !withXAxis,
                dataKey,
                tick: { transform: "translate(0, 10)", fontSize: 12, fill: "currentColor" },
                stroke: "",
                interval: "preserveStartEnd",
                tickLine: withXTickLine ? { stroke: "currentColor" } : false,
                minTickGap: 5,
                ...getStyles("axis"),
                ...xAxisProps,
                children: [
                  xAxisLabel && /* @__PURE__ */ jsx(Label, { position: "insideBottom", offset: -20, fontSize: 12, ...getStyles("axisLabel"), children: xAxisLabel }),
                  xAxisProps?.children
                ]
              }
            ),
            /* @__PURE__ */ jsxs(
              YAxis,
              {
                yAxisId: "left",
                orientation: "left",
                tick: { transform: "translate(-10, 0)", fontSize: 12, fill: "currentColor" },
                hide: !withYAxis,
                ...sharedYAxisProps,
                ...yAxisProps,
                children: [
                  yAxisLabel && /* @__PURE__ */ jsx(
                    Label,
                    {
                      position: "insideLeft",
                      angle: -90,
                      textAnchor: "middle",
                      fontSize: 12,
                      offset: -5,
                      ...getStyles("axisLabel"),
                      children: yAxisLabel
                    }
                  ),
                  yAxisProps?.children
                ]
              }
            ),
            /* @__PURE__ */ jsxs(
              YAxis,
              {
                yAxisId: "right",
                orientation: "right",
                tick: { transform: "translate(10, 0)", fontSize: 12, fill: "currentColor" },
                hide: !withRightYAxis,
                ...sharedYAxisProps,
                ...rightYAxisProps,
                children: [
                  rightYAxisLabel && /* @__PURE__ */ jsx(
                    Label,
                    {
                      position: "insideRight",
                      angle: 90,
                      textAnchor: "middle",
                      fontSize: 12,
                      offset: -5,
                      ...getStyles("axisLabel"),
                      children: rightYAxisLabel
                    }
                  ),
                  yAxisProps?.children
                ]
              }
            ),
            /* @__PURE__ */ jsx(
              CartesianGrid,
              {
                strokeDasharray,
                vertical: gridAxis === "y" || gridAxis === "xy",
                horizontal: gridAxis === "x" || gridAxis === "xy",
                ...getStyles("grid"),
                ...gridProps
              }
            ),
            withTooltip && /* @__PURE__ */ jsx(
              Tooltip,
              {
                animationDuration: tooltipAnimationDuration,
                isAnimationActive: tooltipAnimationDuration !== 0,
                position: { y: 0 },
                cursor: {
                  stroke: "var(--chart-grid-color)",
                  strokeWidth: 1,
                  strokeDasharray
                },
                content: ({ label, payload }) => /* @__PURE__ */ jsx(
                  ChartTooltip,
                  {
                    label,
                    payload,
                    unit,
                    classNames: resolvedClassNames,
                    styles: resolvedStyles,
                    series,
                    valueFormatter
                  }
                ),
                ...tooltipProps
              }
            ),
            lines,
            referenceLinesItems,
            children
          ]
        }
      ) })
    }
  );
});
CompositeChart.displayName = "@mantine/charts/CompositeChart";
CompositeChart.classes = classes;

export { CompositeChart };
